### PR TITLE
Make ExtensionsManager typed

### PR DESCRIPTION
**Pull request by asi1024**
Depends on ~#182, #183, #200 and #205~

### CODE REVIEW

File: pytorch_pfn_extras/training/trigger_util.py [hunk 1]
```diff
@@ -42,10 +59,12 @@ def get_trigger(trigger):
...
     from pytorch_pfn_extras.training.triggers import interval_trigger
 
-    if callable(trigger):
+    if isinstance(trigger, Trigger):
         return trigger
+    elif callable(trigger):
+        return _CallableTrigger(trigger)
     elif trigger is None:
-        return _never_fire_trigger
+        return _CallableTrigger(_never_fire_trigger)
```
Comment (asi1024) [inline pytorch_pfn_extras/training/trigger_util.py:67]: note: This PR fixes `get_trigger` to always return `Trigger` type.

File: pytorch_pfn_extras/training/manager.py [hunk 2]
```diff
@@ -505,25 +550,30 @@ def set_training_started(engine):
...
+            def close_reporter_on_iter(engine: Engine):
                 self.cm.__exit__(None, None, None)
 
         @self.engine.on(Events.COMPLETED)
-        def set_extensions_cleanup(engine):
+        def set_extensions_cleanup(engine: Engine):
             self._finalize_extensions()
 
-    def state_dict(self):
+    def state_dict(self, *, transform_models=None) -> Dict[str, Any]:
```
Comment (asi1024) [inline pytorch_pfn_extras/training/manager.py:566]: `BaseExtensionsManager.state_dict` supports `transform_models` option, but this class did not.

Comment (asi1024) [reply pytorch_pfn_extras/training/manager.py]: Done in 314c2af.

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: pytorch_pfn_extras/training/manager.py [hunk 3]
```diff
@@ -64,7 +71,7 @@ def __init__(
             writer = writing.SimpleWriter(out_dir=out_dir)
         # triggers are stateful, so we need to make a copy for internal use
         self._internal_stop_trigger = copy.deepcopy(self._stop_trigger)
-        self.observation = {}
+        self.observation: Dict[str, torch.Tensor] = {}
```
[the remark this answers is not in the document]
Comment (asi1024) [reply pytorch_pfn_extras/training/manager.py:74]: No. the value might be a scalar. `Dict[str, Union[torch.Tensor, int, float, numpy.generic]]` is correct? Or it must be `Dict[str, Any]`?

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Comment by asi1024** [comment]
A non-`Extension` object seems to be given as an `extension` argument. I will fix it in another PR.

File: pytorch_pfn_extras/training/trigger_util.py [hunk 4]
```diff
@@ -10,7 +13,21 @@ def __call__(self, manager):
         raise NotImplementedError
 
 
-def get_trigger(trigger):
+class _CallableTrigger(Trigger):
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/trigger_util.py:16]: Could you split this to separate PR?

Comment (asi1024) [reply pytorch_pfn_extras/training/trigger_util.py:16]: Split in https://github.com/pfnet/pytorch-pfn-extras/pull/205.

**Comment by kmaehashi** [comment]
Could you resolve conflicts?

**Comment by asi1024** [comment]
Merged master. PTAL.

File: pytorch_pfn_extras/training/manager.py [hunk 5]
```diff
@@ -402,7 +438,11 @@ def __init__(
         self._prepare_for_training(0, iters_per_epoch)
 
     @contextlib.contextmanager
-    def run_iteration(self, *, step_optimizers=None):
+    def run_iteration(
+            self,
+            *,
+            step_optimizers: Optional[List[str]] = None
+    ) -> Iterator[None]:
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:445]: Should it be `typing.Generator`?

File: pytorch_pfn_extras/training/manager.py [hunk 6]
```diff
@@ -454,15 +494,18 @@ class IgniteExtensionsManager(_BaseExtensionsManager):
     """
     def __init__(
             self,
-            engine,
-            models,
-            optimizers,
-            max_epochs,
+            engine: Any,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:497]: Can't we type here?

**Review by kmaehashi — CHANGES_REQUESTED** [review]
LGTM except minor comments 👍🏼

**Comment by asi1024** [comment]
@kmaehashi PTAL.

**Review by kmaehashi — APPROVED** [review]
Thanks, LGTM!
